src/simple_ble.rs

```diff
@@ -32,173 +37,117 @@ pub mod gap_data {
-                self.name.as_bytes(),
+                name.as_bytes(),
             )
-        } {
-            0 => Ok(self),
-            _ => Err(""),
-        }
+        };
+        convert_error_code(result)?;
+        Ok(())
     }
 
-    fn set_advertsing_uuid(&mut self) -> Result<&mut Self, &'static str> {
-        match unsafe {
+    fn set_uuid(uuid: Vec<u16>) -> TockResult<(), isize> {
+        let result = unsafe {
             syscalls::allow16(
                 DRIVER_NUMBER,
                 gap_data::COMPLETE_LIST_16BIT_SERVICE_IDS,
-                &self.uuid,
+                &uuid.to_vec(),
             )
-        } {
-            0 => Ok(self),
-            _ => Err(""),
-        }
+        };
+        convert_error_code(result)?;
+        Ok(())
     }
 
-    fn set_advertising_flags(&mut self) -> Result<&mut Self, &'static str> {
-        match unsafe { syscalls::allow(DRIVER_NUMBER, gap_data::SET_FLAGS, &self.flags) } {
-            0 => Ok(self),
-            _ => Err(""),
-        }
+    fn set_flags(flags: [u8; 1]) -> TockResult<(), isize> {
+        let result = unsafe { syscalls::allow(DRIVER_NUMBER, gap_data::SET_FLAGS, &flags) };
+        convert_error_code(result)?;
+        Ok(())
     }
 
-    fn set_advertising_buffer(&mut self) -> Result<&mut Self, &'static str> {
-        match unsafe {
-            syscalls::allow(
-                DRIVER_NUMBER,
-                ble_commands::ALLOW_ADVERTISMENT_BUFFER,
-                &self.buffer,
-            )
-        } {
-            0 => Ok(self),
-            _ => Err(""),
-        }
+    fn set_service_payload(service_payload: Vec<u8>) -> TockResult<(), isize> {
+        let result =
+            unsafe { syscalls::allow(DRIVER_NUMBER, gap_data::SERVICE_DATA, &service_payload) };
+        convert_error_code(result)?;
+        Ok(())
     }
 
-    fn set_service_payload(&mut self) -> Result<&mut Self, &'static str> {
-        match unsafe {
-            syscalls::allow(DRIVER_NUMBER, gap_data::SERVICE_DATA, self.service_payload)
-        } {
-            0 => Ok(self),
-            _ => Err(""),
-        }
-    }
-}
-
-impl<'a> BleDeviceInitialized<'a> {
-    pub fn start_advertising(&mut self) -> Result<BleDeviceAdvertising, &'static str> {
-        match unsafe { syscalls::command(DRIVER_NUMBER, ble_commands::START_ADVERTISING, 0, 0) } {
-            0 => Ok(BleDeviceAdvertising {
-                interval: &mut self.interval,
-                name: &mut self.name,
-                uuid: &mut self.uuid,
-                flags: &mut self.flags,
-                buffer: &mut self.buffer,
-                service_payload: &mut self.service_payload,
-            }),
-            _ => Err(""),
-        }
+    fn start_advertising() -> TockResult<(), isize> {
+        let result =
+            unsafe { syscalls::command(DRIVER_NUMBER, ble_commands::START_ADVERTISING, 0, 0) };
+        convert_error_code(result)?;
+        Ok(())
```

Comment:
Isn't `foo?; Ok(())` equivalent to `foo`?

Reply:
Right, good catch!